WalletWasabi/Crypto/Api/WabiSabiClient.cs

```diff
@@ -0,0 +1,167 @@
+			}
+
+			foreach (var presented in CredentialsToPresent)
+			{
+				Credentials.Remove(presented.Credential);
+			} 
+			CredentialsToPresent.Clear();
+			CredentialsRequested.Clear();
+			CredentialsToRequest.Clear();
+			var credentialReceived = credentials.Select(x => 
+				new Credential(new Scalar((ulong)x.Requested.Amount.Satoshi), x.Requested.r, x.Issued));
+			Credentials.AddRange(credentialReceived);
+		}
+
+		public ICredentialRequestBuilder AsCredentialRequestBuilder()
+		{
+			return this;
+		} 
+
+		ICredentialRequestBuilder ICredentialRequestBuilder.RequestCredentialFor(Money amount)
+		{
+			Guard.NotNull(nameof(amount), amount);
+			Guard.InRangeAndNotNull(nameof(amount), amount, Money.Zero, Constants.MaximumCredentailAmount);
+
+			CredentialsToRequest.Add(amount);
+
+			return this;
+		}
+
+		ICredentialRequestBuilder ICredentialRequestBuilder.PresentCredentials(params Credential[] credentials)
+		{
+			Guard.NotNullOrEmpty(nameof(credentials), credentials);
+
+			foreach (var credential in credentials)
+			{
+				var z = Random.GetScalar();
+				if (!CredentialsToPresent.Any(x => x.Credential.Mac == credential.Mac))
+				{
+					var knowledge = ProofSystem.ShowCredential(credential.Present(z), z, credential, CoordinatorParameters);
+					CredentialsToPresent.Add((credential, z, knowledge));
+				}
+			}
+			return this;
+		}
+
+		RegistrationRequest ICredentialRequestBuilder.Build()
+		{
+			Knowledge balanceKnowledge;
+
+			foreach (var amount in CredentialsToRequest)
+			{
+				var attribute = Attribute.FromMoney(amount, Random);
+				var scalarAmount = new Scalar((ulong)amount.Satoshi);
+
+				if (IsNullRequest)
+				{
+					var rangeKnowledge = ProofSystem.ZeroProof(attribute.Ma, attribute.Randomness);
+
+					var credentialRequest = new CredentialIssuanceRequest(attribute.Ma, Enumerable.Empty<GroupElement>());
+					CredentialsRequested.Add((amount, credentialRequest, attribute.Randomness, rangeKnowledge));
+				}
+				else
+				{
+					var (rangeKnowledge, bitCommitments) = ProofSystem.RangeProof(scalarAmount, attribute.Randomness, Constants.RangeProofWidth, Random);
+
+					var credentialRequest = new CredentialIssuanceRequest(attribute.Ma, bitCommitments);
+					CredentialsRequested.Add((amount, credentialRequest, attribute.Randomness, rangeKnowledge));
+				}
+			}
+
+			if (IsNullRequest)
+			{
+				var r = Random.GetScalar();
+				var Ma = r * Generators.Gh;
+				balanceKnowledge = ProofSystem.ZeroProof(Ma, r);
+			}
+			else
+			{
+				var balanceData = ComputeBalanceData();
+				balanceKnowledge = ProofSystem.BalanceProof(balanceData.SumOfZ, balanceData.DeltaR);
+			}
+
+			var knowledgeToProve = 	CredentialsToPresent.Select(x => x.Knowledge)
+				.Concat(CredentialsRequested.Select(x => x.Knowledge));
+
+			if (!IsNullRequest)
+			{
+				knowledgeToProve = knowledgeToProve.Concat(new []{ balanceKnowledge });
+			}
+
+			var transcript = new Transcript(new byte[0]);
```

Comment:
IMO transcript should be passed in as an argument by the client

Reply:
I modified the code in order to pass the transcript from the caller but I think this is where the transcript should be created. No higher level api should need to know about the transcript concept.